all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,454 @@
+must work across all SDKs/CLIs.
+
+### Why?
+
+What are the reasons for building this?
+
+* Users want to switch environments without changing their code, but today they cannot.
+* Our samples should be able to be run on multiple environments including cloud, but today they are localhost only.
+
+### Goals
+
+* Common, well-specified configuration format usable by Temporal tooling for client options
+* Implementation in all Temporal SDKs for reading and writing of this configuration
+* All Temporal CLI/SDKs updated to have easy client creation using this configuration
+* Common on-disk location in every platform for reading/writing this format
+* Ability to provide environment variables to represent values in the configuration
+* Well-specified hierarchy for on-disk overridable by environment variables overridable by in-CLI/SDK values
+* 0 new dependencies in SDKs
+* Profile names within the configuration
+
+### Non-goals
+
+* Common format for declarative workers, schedules, etc. This is only for client options.
+* File-based hierarchy. We don't need many levels of file merging for these options at this time.
+* Extreme focus on human authoring of this format. This is because it'd likely go against the "0 new dependencies" rule.
+* Use external configuration by default. Users need to opt-in to this for compatibility and clarity reasons (but it is
+  of course easy).
+
+### High-level Usage Examples
+
+#### Simple Switch from Local to Cloud
+
+Similar to `temporal env` today, you can use the CLI to manage configurations with `temporal config`.
+
+For example, say you had this Go code for starting a workflow:
+
+```go
+options, err := client.LoadOptionsFromConfig("default")
+if err != nil {
+  return err
+}
+cl, err := client.Dial(options)
+run, err := cl.ExecuteWorkflow(
+  ctx,
+  client.StartWorkflowOptions{ID: "my-id", TaskQueue: "my-task-queue"},
+  MyWorkflow,
+)
+```
+
+By default this only works on `localhost:7233` with namespace `default` when you run it. But if you were a cloud user,
+you could do this in environment variables for mTLS auth:
+
+    export TEMPORAL_ADDRESS=my-ns.a1b2c.tmprl.cloud:7233
+    export TEMPORAL_NAMESPACE=my-ns.a1b2c
+    export TEMPORAL_TLS_CLIENT_CERT_PATH=path/to/my/client.pem
+    export TEMPORAL_TLS_CLIENT_KEY_PATH=path/to/my/client.key
+
+Now running the same code again unmodified will run against cloud. You can do ths same thing with config instead of
+environment variables:
+
+    temporal config set --key address --value my-ns.a1b2c.tmprl.cloud:7233
+    temporal config set --key namespace --value my-ns.a1b2c
+    temporal config set --key tls.clientCertPath --value path/to/my/cert.pem
+    temporal config set --key tls.clientKeyPath --value path/to/my/cert.key
+
+Now running the same code again unmodified will run against cloud. Or in a Temporal future with API keys and a
+known/fixed API endpoint:
+
+    temporal config set --key cloudApiKey --value my-api-key
```

Comment:
> yuck, I do not like having cloud and non-cloud config keys

This does not exist today FYI. But it's an example because this is not just a cloud config key, this is saying "default to all the other cloud things".

> I'm wary of even providing "buillt-in" defaults for cloud because I think they will have more confidence/comfort in being able to see exactly what's different.

A single known API domain and that TLS defaults to true are reasonable cloud defaults.

> If we must have different API keys because of talking to the control plane vs. data plane (or whatever)

Again this does not exist today, and that's not why we need different API keys. The reason we want different _config_ values is because there are things you can assume about cloud that you may not be able to assume about others. Regardless, this is just a sample of something we could do in the future. In the meantime, you can use `apiKey` (and always will be able to, even for cloud).

Reply:
> Are you saying that setting cloudApiKey (or any other cloud-specific setting) would hypothetically auto-magically say "you're running against cloud now" and change other config values (e.g. apiKey) automatically?

Yes

> I hate it when libraries do that.

I mean we could have just removed api key and had authentication headers. But we recognize shortcuts are important for ergonomics. We may in the future be able to accept a single value from a user to run on cloud (whether in the CLI, SDK, config, etc) and we should consider it.